stdlib/either.mli

```diff
@@ -0,0 +1,88 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*         Gabriel Scherer, projet Parsifal, INRIA Saclay                 *)
+(*                                                                        *)
+(*   Copyright 2019 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Either type.
+
+    Either is the simplest and most generic sum type. It is isomorphic
+    to the [result] type; it should be used when there is no implied
+    success/failure semantics in the choice of one of the constructors.
+
+    @since 4.11 *)
+
```

Comment:
The introduction gives the impression that `result` should be the default, and `Either.t` a kind of fall-back when `result` won't do, which sounds a bit apologetic.  (Similarly below, in the comment starting "Unlike [result]")

Reply:
I think that it makes sense to compare to the type that was introduced earlier.

I tried to give a more positive description of when either should be used, and I had the idea the idea of saying that it makes sense for "symmetric" sums, but actually this is not correct: most use-cases of either will give very asymmetric meanings to the two cases, it's just that the mapping is not a-priori directed by the idea that there is a "good" and "bad" case.

One thing I could say is that either is a primitive sum type whose constructor names are generic/interchangeable, and *then* point out that for values that encode a form of  failure `result` is more precise?

Reply:
> I think that it makes sense to compare to the type that was introduced earlier.

I actually disagree with this.  Most OCaml users are in the future, and (if this is merged) will encounter a standard library that already contains both `either` and `result`; there's no reason to present them with documentation that depends on the order in which the types were introduced.

On the other hand, I think it's a good idea to have your justification for `either` in the pull request and the commit message.

Reply:
I'll try to rewrite this without a reference to `result`.